Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm

```diff
@@ -279,12 +283,24 @@ static void didReceiveServerTrustChallenge(NetworkConnectionToWebProcess& connec
         case nw_connection_group_state_waiting:
             return; // We will get another callback with another state change.
         case nw_connection_group_state_ready:
+            if (RefPtr protectedThis = weakThis.get())
+                protectedThis->m_sessionMetadata = nw_connection_group_copy_protocol_metadata(protectedThis->m_connectionGroup.get(), adoptNS(nw_protocol_copy_webtransport_definition()).get());
             return creationCompletionHandler(true);
         case nw_connection_group_state_failed:
-            creationCompletionHandler(false);
-            if (RefPtr protectedThis = weakThis.get())
-                protectedThis->send(Messages::WebTransportSession::DidFail());
-            return;
+            if (RefPtr protectedThis = weakThis.get()) {
+                if (RetainPtr metadata = protectedThis->m_sessionMetadata) {
+                    std::optional<unsigned> sessionErrorCode = std::nullopt;
+                    String sessionErrorMessage;
+                    if (canLoad_Network_nw_webtransport_metadata_get_session_closed() && softLink_Network_nw_webtransport_metadata_get_session_closed(metadata.get())) {
+                        sessionErrorCode = nw_webtransport_metadata_get_session_error_code(metadata.get());
+                        if (const char* errorMessage = nw_webtransport_metadata_get_session_error_message(metadata.get()))
+                            sessionErrorMessage = String::fromUTF8(unsafeSpan(errorMessage));
+                    }
+                    protectedThis->send(Messages::WebTransportSession::DidFail(WTFMove(sessionErrorCode), WTFMove(sessionErrorMessage)));
+                    return;
```

Comment:
This should probably be removed, so we call creationCompletionHandler in all failure cases, right?

Reply:
Return will execute only when we have metadata, and metadata is present only when connection group reaches ready state. When connection group reaches ready state, completion handler is already called.
So removing return will be redundant in this case.